src/rez/resolved_context.py

```diff
@@ -162,16 +177,28 @@ def __call__(self, state):
                 return self.callback(state)
             return SolverCallbackReturn.keep_going, ''
 
-    def __init__(self, package_requests, verbosity=0, timestamp=None,
-                 building=False, caching=None, package_paths=None,
-                 package_filter=None, package_orderers=None, max_fails=-1,
-                 add_implicit_packages=True, time_limit=-1, callback=None,
-                 package_load_callback=None, buf=None, suppress_passive=False,
-                 print_stats=False, package_caching=None):
+    def __init__(self,
+                 package_requests: Iterable[str | Requirement],
+                 verbosity=0,
+                 timestamp: float | None = None,
+                 building=False,
+                 caching: bool | None = None,
+                 package_paths: list[str] | None = None,
+                 package_filter: PackageFilterList | None = None,
+                 package_orderers: list[PackageOrder] | None = None,
+                 max_fails=-1,
+                 add_implicit_packages=True,
+                 time_limit=-1,
+                 callback: Callable[[SolverState], tuple[SolverCallbackReturn, str]] | None = None,
+                 package_load_callback: Callable[[Package], Any] | None = None,
+                 buf: SupportsWrite | None = None,
+                 suppress_passive=False,
+                 print_stats=False,
+                 package_caching=None):
         """Perform a package resolve, and store the result.
 
         Args:
-            package_requests (list[typing.Union[str, PackageRequest]]): request
+            package_requests (list[typing.Union[str, Requirement]]): request
```

Comment:
I noticed that everywhere that we've documented types as `PackageRequest`, they appear to actually be `Requirement`.  I'm not sure if there any real-world exceptions to this.

Reply:
The kind of haphazard use and documentation of `PackageRequest` and `PackageRequest` results in some very difficult situations to accurate add type annotations.   If you want to see for yourself, check out the code, change this to `PackageRequest` and observe the new errors produced by mypy.